connexion/security/security_handler_factory.py

```diff
@@ -0,0 +1,328 @@
+
+    @staticmethod
+    def get_cookie_value(cookies, name):
+        '''
+        Returns cookie value by its name. None if no such value.
+        :param cookies: str: cookies raw data
+        :param name: str: cookies key
+        '''
+        cookie_parser = http_cookies.SimpleCookie()
+        cookie_parser.load(str(cookies))
+        try:
+            return cookie_parser[name].value
+        except KeyError:
+            return None
+
+    def verify_api_key(self, api_key_info_func, loc, name):
+        check_api_key_func = self.check_api_key(api_key_info_func)
+
+        def wrapper(request, required_scopes):
+            if loc == 'query':
+                api_key = request.query.get(name)
+            elif loc == 'header':
+                api_key = request.headers.get(name)
+            elif loc == 'cookie':
+                cookie_list = request.headers.get('Cookie')
+                api_key = self.get_cookie_value(cookie_list, name)
+            else:
+                return self.no_value
+
+            if api_key is None:
+                return self.no_value
+
+            return check_api_key_func(request, api_key, required_scopes=required_scopes)
+
+        return wrapper
+
+    def verify_bearer(self, token_info_func):
+        """
+        :param token_info_func: types.FunctionType
+        :rtype: types.FunctionType
+        """
+        check_bearer_func = self.check_bearer_token(token_info_func)
+
+        def wrapper(request, required_scopes):
+            auth_type, token = self.get_auth_header_value(request)
+            if auth_type != 'bearer':
+                return self.no_value
+            return check_bearer_func(request, token, required_scopes=required_scopes)
+
+        return wrapper
+
+    def _need_to_add_context_or_scopes(self, func):
+        arguments, has_kwargs = inspect_function_arguments(func)
+        need_context = self.pass_context_arg_name and (has_kwargs or self.pass_context_arg_name in arguments)
+        need_required_scopes = has_kwargs or self.required_scopes_kw in arguments
+        return need_context, need_required_scopes
+
+    def _generic_check(self, func, exception_msg):
+        need_to_add_context, need_to_add_required_scopes = self._need_to_add_context_or_scopes(func)
+
+        def wrapper(request, *args, **kw):
+            kwargs = {}
+            if need_to_add_context:
+                kwargs[self.pass_context_arg_name] = request.context
+            if need_to_add_required_scopes:
+                # required_scopes is passed in kw because python 2.7 does not support syntax:
+                # def wrapper(request, *args, required_scopes=None)
+                kwargs[self.required_scopes_kw] = kw.get(self.required_scopes_kw)
+            token_info = func(*args, **kwargs)
+            if token_info is self.no_value:
+                return self.no_value
+            if token_info is None:
+                raise OAuthResponseProblem(description=exception_msg)
```

Comment:
This is odd. Here we're using `OAuthResponseProblem` for bearer token, basic auth and api key, not just for oauth. look for all of the `_generic_check` usage to see what I'm talking about.